torchinductor/scheduler.py

```diff
@@ -1000,188 +761,279 @@ def dead_node_elimination(self):
+            return False
+        if isinstance(node1, (ExternKernelSchedulerNode, NopKernelSchedulerNode)):
+            return False
+        if isinstance(node2, (ExternKernelSchedulerNode, NopKernelSchedulerNode)):
+            return False
+        if node2.get_names() & node1.recursive_predecessors:
+            return False  # node2 must go before node1
+        if node2.is_template():
+            return False  # only epilogues
+
+        device = node1.get_device()
+        if device != node2.get_device():
+            return False  # wrong device
+
+        no_shared_data = self.score_fusion_memory(node1, node2) == 0
+        if no_shared_data and (
+            not config.aggressive_fusion or node1.is_reduction() or node2.is_reduction()
+        ):
+            return False  # heuristic not needed for correctness
+
+        if node1.get_names() & node2.recursive_predecessors:
+            # node2 depends on node1 outputs
+            if not self.can_fuse_vertical(node1, node2):
+                return False
+            if node1.is_template():
+                return template_can_fuse(node1, node2)
+            return self.get_backend(device).can_fuse_vertical(node1, node2)
+        else:  # nodes don't depend on each other, but may have common reads
+            if node1.is_template():
+                return False
+            return self.get_backend(device).can_fuse_horizontal(node1, node2)
+
+    def can_fuse_vertical(self, node1, node2):
+        """
+        Check if it is legal to fuse a consumer (node2) into a producer (node1).
+
+        We can fuse them if all the reads of node2 either match
+        corresponding writes in node1, or are written by nodes that can
+        be scheduled before the fusion of node1 and node2.
+        """
+        node1_names = node1.get_names()
+        remaining_deps = {
+            dep.name for dep in node2.unmet_dependencies - node1.read_writes.writes
+        }
+        if remaining_deps & node1_names:
+            return False  # MemoryDeps didn't match
+        for name in remaining_deps:
+            if node1_names & self.name_to_fused_node[name].recursive_predecessors:
+                return False
+        return True
+
+    def score_fusion(self, node1: BaseSchedulerNode, node2: BaseSchedulerNode):
+        """
+        Assign a score (higher comes first) to the fusion of node1
+        and node2.  When different fusions conflict with each other,
+        this is the way we decide what order to run them in.
+
+        Our current score is based on:
+        - Estimate of the saved memory operations
+        - Fusions closer together in original order
+        """
+        memory_score = self.score_fusion_memory(node1, node1)
+        proximity_score = -max(
+            abs(node1.min_order - node2.max_order),
+            abs(node2.min_order - node1.max_order),
+        )
+        return (
+            not node1.is_reduction() and not node2.is_reduction() and memory_score > 0,
```

Comment:
Hrmm... do we want to fuse the pointwise ops first?

Reply:
Running pointwise fusions first is needed to fix those regressions on internal models.  I'm not totally sure why.  I'd like to here @ngimel's thoughts.

Reply:
We should definitely come back to this heuristic in the future and try as many thing we can think of.  I only tried 4 things then stopped.